maskinporten_api/audit.py

```diff
@@ -0,0 +1,40 @@
+import logging
+from datetime import datetime, timezone
+
+import boto3
+from botocore.exceptions import ClientError
+
+log = logging.getLogger()
+
+
+def audit_log(item_id, item_type, env, action, user, scopes=None, client_id=None):
```

Comment:
Jeg ville kanskje definert en dataklasse istedenfor å ha så mange argumenter i metoden 
```
@dataclass
class AuditLogEntry:
    item_id: str
    item_type: str
    env: str
    action: str # 
    user: str
    scopes = {}
    client_id = {}
```
Og så synes jeg at det kanskje burde være en klasse 
```
class AuditLogger:

    def __init__(self):
        dynamodb = boto3.resource("dynamodb", region_name="eu-west-1")
        self.table = dynamodb.Table("maskinporten-audit-trail")

    def log(self, audit_log_entry: AuditLogEntry):
        """Add a log entry to the API's audit trail."""
```
   
Så kan den initialiseres i maskinporten_clients.py `audit_logger = AuditLogger()`, da kalles `boto3.resource("dynamodb", region_name="eu-west-1")` og `dynamodb.Table("maskinporten-audit-trail")` kun én gang per oppstart og så blir syntaksen for å logge `audit_logger.log()` litt mer expressive i mine øyne

Reply:
> Jeg ville kanskje definert en dataklasse istedenfor å ha så mange argumenter i metoden
> 
> ```
> @dataclass
> class AuditLogEntry:
>     id: str
>     type: str
>     env: str
>     action: str # 
>     user: str
>     scopes = {}
>     client_id = {}
> ```

Hm, men litt rart/un-Pythonish å måtte importere en ekstra klasse for å kunne gjøre funksjonskallet? Lurte på å gjøre det til en dictionary i stedet for å ha så mange parametere, men ulempen da er man man mister den gratis valideringen. Helte mot å ha parameterne i stedet da, siden det er ganske vanlig å ha mange (keyword) parametere ellers i Python.

> Og så synes jeg at det kanskje burde være en klasse
> 
> ```
> class AuditLogger:
> 
>     def __init__(self):
>         dynamodb = boto3.resource("dynamodb", region_name="eu-west-1")
>         self.table = dynamodb.Table("maskinporten-audit-trail")
> 
>     def log(self, audit_log_entry: AuditLogEntry):
>         """Add a log entry to the API's audit trail."""
> ```
> 
> Så kan den initialiseres i maskinporten_clients.py `audit_logger = AuditLogger()`, da kalles `boto3.resource("dynamodb", region_name="eu-west-1")` og `dynamodb.Table("maskinporten-audit-trail")` kun én gang per oppstart og så blir syntaksen for å logge `audit_logger.log()` litt mer expressive i mine øyne

Kan ha det som en klasse, men prøver å få oss bort fra å initialisere objekter på modulnivå. Når man kjører kode som har side-effekter ved import-time kan det skje uventa ting (som de boto-testene som begynner å feile), så ville fortsatt initialisert det per kall, da.